pkg/plugins/manager_test.go

```diff
@@ -0,0 +1,145 @@
+package plugins
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"strconv"
+	"testing"
+
+	"github.com/hashicorp/go-plugin"
+)
+
+const registeredPluginName = "plugin"
+
+var basicHS = PluginHandshake{
+	Key:   "key",
+	Value: "value",
+}
+
+type testCase struct {
+	m           *Manager[PingPongStub]
+	name        string
+	p           plugin.Plugin
+	expectedErr error
+	description string
+}
+
+type testCases []testCase
+
+func TestManager_LoadPluginClient(t *testing.T) {
+	id := PluginIdentity{
+		ProtocolVersion: 1,
+	}
+	cases := testCases{
+		{
+			NewManager[PingPongStub](),
+			registeredPluginName,
+			GRPCPlugin{Impl: &PingPongPlayer{}},
+			nil,
+			"successful registration",
+		},
+		{
+			&Manager[PingPongStub]{},
+			registeredPluginName,
+			GRPCPlugin{Impl: &PingPongPlayer{}},
+			ErrUninitializedManager,
+			"uninitialized manager",
+		},
+		{
+			NewManager[PingPongStub](),
+			"someNonExistingPlugin",
+			GRPCPlugin{Impl: &PingPongPlayer{}},
+			ErrPluginNameNotFound,
+			"unregistered service",
+		},
+		{
+			NewManager[PingPongStub](),
+			registeredPluginName,
+			NopGRPCPlugin{},
+			ErrPluginOfWrongType,
+			"wrong type plugin",
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.description, func(t *testing.T) {
+			// Necessary to redefine in every test due to future setting of stdout by the go-plugin package
+			id.Cmd = *pluginServerCmd(id.ProtocolVersion, basicHS)
+			if tc.p != nil {
+				tc.m.RegisterPlugin(registeredPluginName, id, basicHS, tc.p)
+			}
+			_, err := tc.m.LoadPluginClient(tc.name)
+			assertErr(t, err, tc.expectedErr, tc.description)
+		})
+	}
+}
+
+func assertErr(t *testing.T, err, cmpErr error, desc string) {
+	if err != nil && !errors.Is(err, cmpErr) {
+		t.Errorf("'%s' failed: %v", desc, err)
+	}
+}
+
+// Used to run the plugin server
+func pluginServerCmd(version uint, auth PluginHandshake) *exec.Cmd {
+	cs := []string{"-test.run=TestPluginServer", "--"}
+	cs = append(cs, auth.Key, auth.Value, strconv.Itoa(int(version)))
+	cmd := exec.Command(os.Args[0], cs...)
+
+	env := []string{
+		"RUN_PLUGIN_SERVER=1",
+	}
+	cmd.Env = append(env, os.Environ()...)
+	return cmd
+}
+
+// This is not a real test. This is the plugin server that will be triggered by the tests
+func TestPluginServer(*testing.T) {
+	if os.Getenv("RUN_PLUGIN_SERVER") != "1" {
+		return
+	}
+
+	defer os.Exit(0)
```

Comment:
Not sure why a test helper should ever exit.  This is even worse, it exits successfully when something strange happens.  But if that happens, the test succeeds _even though it should fail_.

Reply:
true this is useless